modules/starfish/2.0/starfish2.smk

```diff
@@ -0,0 +1,247 @@
+    conda: 
+        CFG["conda_envs"]["bcftools"]
+    threads:
+        CFG["threads"]["starfish_union"]
+    resources:
+        **CFG["resources"]["starfish_union"]
+    shell: 
+        op.as_one_line("""
+        bcftools merge {params.options} --force-samples 
+        -Oz -o {output.union} 
+        {input} && 
+        tabix -p vcf {output.union}
+        """)
+
+# Symlinks the final output files into the module results directory (under '99-outputs/')
+rule _starfish_output_vcf:
+    input:
+        vcf = CFG['dirs']['starfish'] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{vcf_name}.vcf.gz"
+    output:
+        vcf = CFG["dirs"]["outputs"] + "vcf/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.{vcf_name}.vcf.gz", 
+        tbi = CFG["dirs"]["outputs"] + "vcf/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.{vcf_name}.vcf.gz.tbi"
+    wildcard_constraints: 
+        vcf_name = "\S+(?<!union)"
+    run:
+        op.relative_symlink(input.vcf, output.vcf), 
+        op.relative_symlink(input.vcf + ".tbi", output.tbi)
+
+rule _starfish_output_union:
+    input:
+        vcf = CFG['dirs']['starfish'] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.vcf.gz", 
+        tbi = CFG['dirs']['starfish'] + "{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}/{union_vcf}.vcf.gz.tbi",
+    output:
+        vcf = CFG["dirs"]["outputs"] + "vcf/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.{union_vcf}.vcf.gz", 
+        tbi = CFG["dirs"]["outputs"] + "vcf/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.{union_vcf}.vcf.gz.tbi"
+    wildcard_constraints: 
+        union_vcf = union_vcf
+    run:
+        op.relative_symlink(input.vcf, output.vcf), 
+        op.relative_symlink(input.tbi, output.tbi)
+
+rule _starfish_output_venn: 
+    input: 
+        venn = str(rules._starfish_run.output.venn)
+    output: 
+        venn = CFG["dirs"]["outputs"] + "venn/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.venn.pdf"
+    run: 
+        op.relative_symlink(input.venn, output.venn)
+
+def _starfish_get_output_target(wildcards): 
+    # Get the checkpoint output directory
+    CFG = config["lcr-modules"]["starfish"]
+    checkpoint_output = os.path.dirname(checkpoints._starfish_rename_output.get(**wildcards).output.renamed)
+    # Rename each VCF with the variant caller names
+    vcfs = expand(
+        CFG["dirs"]["outputs"] + "vcf/{seq_type}--{genome_build}/{tumour_id}--{normal_id}--{pair_status}.{vcf_name}.vcf.gz",
+        seq_type = wildcards.seq_type,
+        genome_build = wildcards.genome_build, 
+        tumour_id = wildcards.tumour_id, 
+        normal_id = wildcards.normal_id, 
+        pair_status = wildcards.pair_status, 
+        vcf_name = glob_wildcards(os.path.join(checkpoint_output, "{vcf_name}.vcf.gz")).vcf_name
+        )
+    return vcfs
+
+
+rule _starfish_dispatch: 
```

Comment:
Why (or do) we need a dispatch if the existence of the union file depends on all the others completing? I am not saying it should change but I'm just wondering.

Reply:
It was the only way to generate the symlinks once per VCF.